Prioritization of Tap in Dependency Resolution

CNA-Bld opened the pull request:
This pr implements tap prioritization in dependency resolution. While resolving dependency for a formula `user/repo/foobar`, it will try to look for its dependencies in tap `user/repo` first. If it is not found, it will then use `foobar` formula from core. It will never use dependency formulae from taps other than `user/repo`.

This behavior is suggested by @jacknagel [here](https://github.com/Homebrew/homebrew/issues/14089#issuecomment-39590767).

~~This pr has a little dependency on #41736 as it requires the new behavior of `Formulary#factory` about fully qualified names like `Homebrew/homebrew/foobar`.~~ This behavior is already merged separately in #42550.

xu-cheng commented on Library/Homebrew/dependency_collector.rb:
```diff
@@ -84,11 +85,11 @@ def parse_string_spec(spec, tags)
       TapDependency.new(spec, tags)
     elsif tags.empty?
-      Dependency.new(spec, tags)
+      PrioritizedTapDependency.new("#{@tap}/#{spec}", tags)
```
Instead of string concat, you can just pass tap as another argument. i.e. `PrioritizedTapDependency.new(@tap, spec, tags)`. Of course you need to change the construction method accordingly.

xu-cheng commented on Library/Homebrew/dependency.rb:
```diff
@@ -145,3 +145,21 @@ def installed?
+      Formulary.factory(name)
+    rescue FormulaUnavailableError
+      @tap = "Homebrew/homebrew"
+    end
```
I think it's better to move this logic to `to_formula`. That is only resolving to formula when necessary.

Personally, I think it should be allowed to load formula from other taps(maybe showing a warning in the mean time). Therefore, I think you can take such implementation:

```
class PrioritizedTapDependency < Dependency
  attr_reader :tap

  def initialize(name, tap, tags=[], env_proc=DEFAULT_ENV_PROC, option_name=name)
    @tap = tap
    super name, tags, env_proc, option_name
  end

  def to_formula
    begin
      formula = Formulary.factory("#{tap}/#{name}")
    rescue FormulaUnavailableError
      formula = Formulary.factory(name)
      unless formula.core_formula?
        opoo "<warning message>"
      end
    end
    formula.build = BuildOptions.new(options, formula.options)
    formula
  end
end
```

xu-cheng commented:
Good work here.

xu-cheng commented on Library/Homebrew/software_spec.rb:
```diff
@@ -47,6 +47,7 @@ def initialize
   def owner= owner
     @name = owner.name
     @full_name = owner.full_name
+    @dependency_collector.tap = owner.tap
```
You need to be aware that the tap here is not `Tap` class but result of  `Formula#tap`. It can return following results:
* `nil`: unknown origin we should treat it as core.
* `"Homebrew/homebrew"`: core.
* `"<user>/homebrew-<repo>"`: tap. You probably want to remove the `homebrew-`

It can be a little messy here. If you want, you can try to clean them.(e.g. turn them to `Tap` object)

CNA-Bld commented:
@xu-cheng Thanks for the response! I think allowing dependency resolution across taps might lead to some confusing situations, like:

* We have a formula `repo1/foobar` and it depends on `a` and `b`.
* `repo1` provides `a`, which depends on `c`.
* `repo1` does not provide `b` but `repo2` provides, and it also depends on `c`.
* both `repo1` and `repo2` provide `c`.

So the dependency tree is:

```
repo1/foobar --> repo1/a --> repo1/c
             |-> repo2/b --> repo2/c
```

In this case, we won't be able to choose the proper `c` to install. Shall we have a discussion on whether to allow this?

xu-cheng commented:
@CNA-Bld the case you presented won't work in either allowing across tap dependency resolution or not. The reason I think we should support it is because following common scenario. Say a formula from 3rd party tap depends on a core formula, and for some reason we decide to migrate that core formula to one of official taps. If we allow across tap dependency resolution plus homebrew will automatically tap that tap during update, everything will work fine. However, if we don't allow such behavior, that formula is simply broken until formula author updates it to full-qualified name.

CNA-Bld commented:
@xu-cheng I see. Today I read more on our dependency resolution logic. I think we can add some warning message / raise some error in `Dependency#merge_repeats` when such situation happens.

xu-cheng commented:
If the case you mentioned, homebrew will raise error from `FormulaLock`. So you don't need to worry about it.

mikemcquaid commented:
@Homebrew/owners I'd like to try and get this merged soon (this week or next week) so let us known if you have any more feedback.

xu-cheng commented:
@mikemcquaid Just a reminder that this is not the GSoC PR.

mikemcquaid commented:
@xu-cheng Sure but I'd like to see at least one PR merged per student and there's only two weeks left of the GSoC.

xu-cheng commented on Library/Homebrew/dependency.rb:
```diff
@@ -145,3 +145,31 @@ def installed?
+  attr_reader :tap
+
+  def initialize(name, tags=[], env_proc=DEFAULT_ENV_PROC, option_name=name)
+    @tap, _, option_name = option_name.rpartition "/"
```
Let's make tap an input argument rather than doing string manipulation here.

CNA-Bld replied:
If we don't keep the constructor the same as `Dependency` we need to create a new version of `Dependency#merge_repeats` since it constructs `Dependency` instances there.

xu-cheng replied:
I see the problem here. To simplify it, I think we can do something like this:
* Let's use `owner_tap` instead of `tap` as the attribute name, so it won't be confused with `tap` attribute from `TapDependency`.
* Let's make `owner_tap` as a common attributes for all `Dependency` class. So you only need to update the `Dependency.merge_repeats` method instead of creating a new one.
* I don't think we need another class of `PrioritizedTapDependency`. You can just make necessary change in `Dependency` class itself, such as:
```ruby
class Dependency

  def initialize(name, tags = [], env_proc = DEFAULT_ENV_PROC, option_name = name, owner_tap = nil) 
    ...
  end

  def to_formula
    if owner_tap.nil? || name ~= HOMEBREW_TAP_FORMULA_REGEX
      formula = Formulary.factory(name)
    else
      begin
        formula = Formulary.factory("#{owner_tap}/#{name}")
      rescue FormulaUnavailableError
        formula = Formulary.factory(name)
        unless formula.core_formula?
          opoo "<warning message>"
        end
      end
    end
    formula.build = BuildOptions.new(options, formula.options)
    formula
  end
end
```

In this approach, if owner_tap is nil, it means we don't care the where formula is initialize from. What do you think?

xu-cheng commented on Library/Homebrew/dependency.rb:
```diff
@@ -145,3 +145,31 @@ def installed?
+          You may wish to contact the author of #{@tap} to report this.
+        EOS
+      end
+    end
```
Move the whole block to `to_formula` please.

CNA-Bld replied:
Somehow it does not work if I move it... I think I have some misunderstanding about dependency resolve procedure and I am looking for the problem. Sorry for that :(

xu-cheng commented on Library/Homebrew/dependency_collector.rb:
```diff
@@ -34,6 +34,14 @@ def initialize
+    else
+      @tap = "Homebrew/homebrew"
+    end
+  end
```
To be honestly, it can be strange to have the method here. Maybe put it inside `PrioritizedTapDependency`.

xu-cheng commented:
I just noticed that there is another place you may need to prioritized: that is `Requirement#default_formula`. Although, we never encountered the situation when `default_formula` is a tap formula.

xu-cheng commented:
@CNA-Bld Are you still working on this?

CNA-Bld commented:
@xu-cheng Yes I am. I am too busy this week for some family affairs (which I have to settle by this week since I am going back to Singapore on Sunday). But I will be able to continue on next week. Sorry for that :(

xu-cheng commented on Library/Homebrew/dependency.rb:
```diff
@@ -29,7 +36,23 @@ def hash
+        formula = Formulary.factory("#{owner_tap}/#{name}")
+      rescue FormulaUnavailableError
+        formula = Formulary.factory(name)
+        unless formula.core_formula? or WARNED_CROSS_TAP_DEPENDENCY.include?(name)
```
I suggest to use a global variable here like `@@warned_cross_tap_dependency`

xu-cheng commented on Library/Homebrew/dependency.rb:
```diff
@@ -125,7 +148,14 @@ def merge_repeats(all)
         tags = deps.flat_map(&:tags).uniq
-        dep.class.new(name, tags, dep.env_proc)
+        taps = deps.map(&:owner_tap).flatten.uniq
+        unless taps.length == 1
```
You may want to remove `nil` tap here. I think it's fine to have a nil tap and another tap together. Right?

xu-cheng commented on Library/Homebrew/dependency.rb:
```diff
@@ -4,15 +4,22 @@
+      @owner_tap = "#{$1}/#{$3.gsub(/^homebrew-/, "")}"
+    else
+      @owner_tap = nil
+    end
```
In fact, you can just make it `@owner_tap = owner_tap` here. The reason is `Formulary.factory` can accept ref likes `homebrew/homebrew-foo/bar`.

xu-cheng commented on Library/Homebrew/formula.rb:
```diff
@@ -1256,7 +1257,11 @@ def go_resource(name, &block)
+    end
+
+    def deps
+      @deps ||= []
```
Mind to elaborate a bit why do you need these changes here?

xu-cheng commented:
You need rebase after I merge #43107. After that you will be able to prioritize tap in Requirement.

Considering the profound changes here, I strongly suggest you to create new test cases to prevent it broken in the future.

Good work here. :+1: :clap:

CNA-Bld commented:
@xu-cheng Thanks a lot (again) for code review!

Yesterday I thought I found a bug causing infinite loop which I cannot reproduce on my friend's laptop. After careful tracing I realized that it was just because `to_formula` is taking too much time. It is called multiple times during dependency resolution so I add a cache for it in the recent commit.

Just now I noticed that `Formula#tap` does actually return `Homebrew/homebrew` for core formulas so I added the logic to `Dependency#initialize` to replace it with `nil`.

As for the changes in `formula.rb`, `SoftwareSpec.owner=` is not called yet when we define a formula using the formula DSL (i.e., when `depends_on` is called), thus it is not possible for `DependencyCollector` to get the formula's tap information at the time. That is why I delayed the dependency resolution until `set_spec` is called in `initialize`. I'm not sure if this is the best practice to do this. Can you suggest?

xu-cheng commented:
> As for the changes in formula.rb, SoftwareSpec.owner= is not called yet when we define a formula using the formula DSL (i.e., when depends_on is called), thus it is not possible for DependencyCollector to get the formula's tap information at the time. That is why I delayed the dependency resolution until set_spec is called in initialize. 

In the case, `depends_on` inside stable devel and head block will fail to get tap information as well.

xu-cheng commented on Library/Homebrew/dependency.rb:
```diff
@@ -123,9 +143,16 @@ def merge_repeats(all)
+        dep = deps.first
         tags = deps.flat_map(&:tags).uniq
-        dep.class.new(name, tags, dep.env_proc)
+        taps = deps.map(&:owner_tap).flatten.uniq.select { |tap| not tap.nil? }
```
`taps = deps.flat_map(&:owner_tap).uniq.compat`

xu-cheng commented:
Sorry being busy recently. I think we can solve current problem in this way:

Instead of set `owner_tap` during `initialize`. We can set it lately during `SoftwareSpec.owner=` as such:
```ruby
def owner=(owner)
...
@dependency_collector.deps.each { |dep| dep.owner_tap = owner.tap }
@dependency_collector.requirements.each { |req| req.owner_tap = owner.tap }
...
end
```

CNA-Bld commented:
@xu-cheng Sorry I was extremely busy since the semester started. I may only be able to work on this for a very limited time.

I have implemented it using the way you suggested and cleaned the history a little bit.

xu-cheng commented:
> Sorry I was extremely busy since the semester started. I may only be able to work on this for a very limited time.

If that's the case, I can pick up your work and work on it afterwards.

xu-cheng commented:
Continuing the work at #43535 

Thanks for your contributions @CNA-Bld
